conbench/api/history.py

```diff
@@ -42,10 +53,146 @@ def get(self, benchmark_result_id):
+
+        csv_buf = generate_csv_history_for_result(benchmark_result_id, items)
+
+        return send_file(
+            csv_buf,
+            as_attachment=True,
+            # Use a "timeseries" fingerprint here that represents what all data
+            # points in this series have in common: benchmark name, case perm,
+            # hardware, context, repo.
+            download_name=f"conbench-history-{items[0].benchmark_name}-{items[0].ts_fingerprint}.csv",
+            mimetype="text/csv",
+        )
+
+        # # Note: wrap this into an array if there is just 1 sample, for
+        # # consistency (clients expect an array).
+        # jsonbytes: bytes = orjson.dumps(
+        #     [s._dict_for_api_json() for s in samples],
+        #     option=orjson.OPT_INDENT_2,
+        # )
+        # return csv_response_for_byte_sequence(csv_buf.getbuffer(), 200)
+
+
 history_entity_view = HistoryEntityAPI.as_view("history")
+history_download_endpoint = HistoryDownloadAPI.as_view("history-download")
+
+rule(
+    "/history/download/<benchmark_result_id>/",
+    view_func=history_download_endpoint,
+    methods=["GET"],
+)
+
 
 rule(
     "/history/<benchmark_result_id>/",
     view_func=history_entity_view,
     methods=["GET"],
 )
+
+# deduplicate: move to conbench.numstr
+numstr8 = functools.partial(conbench.numstr.numstr, sigfigs=8)
+
+
+def generate_csv_history_for_result(
+    input_result_id: str, items: List[HistorySample]
+) -> BytesIO:
+    """
+    input from `get_history_for_benchmark()`. Intrinsics of that function
+    matter a lot; read its docstring and code. For example, two important
+    aspects about that func:
+    - results for default branch only
+    - order is not guaranteed in returned collection.
+    """
+
+    assert len(items) > 0
+
+    # Note that this might start to have similarities to the dataframe aspects
+    # within entities.history.execute_history_query_get_dataframe() -- that's
+    # expected. let's streamline internal and external interface in the future
+    # and then we can make analysis on these dataframes (easy-ish) testable.
+    df = pd.DataFrame(
+        # Note(jp:): cannot use a generator expression here, len needs
+        # to be known.
+        {
+            "result_id": [i.benchmark_result_id for i in items],
+            "commit_hash": [i.commit_hash for i in items],
+            "svs": [i.svs for i in items],
+            "min": [min(i.data) for i in items],
+        },
+        # Note(jp): also no generator expression possible. The
+        # `unit="s"` is the critical ingredient to convert this list of
+        # floaty unix timestamps to datetime representation. `utc=True`
+        # is required to localize the pandas DateTimeIndex to UTC
+        # (input is tz-naive).
+        index=pd.to_datetime([i.commit_timestamp for i in items], utc=True),
+    )
+
+    # Sort by time. old -> new
+    df = df.sort_index()
+    df.index.rename("commit_time", inplace=True)
```

Comment:
Validating my reading: In the CSV, this results in the index looking just like any other column, correct? Hopefully?

Reply:
Exactly!

This is super intermediate (I trust we want to change a lot here over time, maybe even in this PR), but so far the head of the generated CSV document looks like
```
# original URL: http://127.0.0.1:5000/api/history/download/0645e82e9bbe7b1b80004c6f2a731cba
# generated by conbench, commit 1748f11628aac62f9213046dbbfa88f8212e467e
# generated at 2023-07-25T12:53:33+00:00
# for result 0645e82e9bbe7b1b80004c6f2a731cba
# benchmark name: tpch-snip
# case permutation: dataset_subvariant=default dataset_variant=parquet-float64 snip ptds=False query_id=snip-09-edited-snip scale_factor=1000 transport_protocol=snip
# hardware hash: snip-infra-748d68f991c92244cefcea963cf3f74f
# timeseries fingerprint: 47c516cec101661312f3fb65eaa90d7c
# single value summary (SVS) type: mean
commit_time,result_id,commit_hash,svs,min
2023-05-03 19:20:30+00:00,06452cc6a52579608000d427b742c4c1,7c4a4fb8f78c55cd072d330040f4e8c775bceaa9,48.666006,48.666006
2023-05-04 00:20:08+00:00,06453129610473c480002253378ce1e3,ade9fd3861cf20d34e4b498ef1c5f573a1a8e5eb,50.470924,50.470924
2023-05-04 02:25:54+00:00,064532fa0b3177968000a11fbb525820,a536476ca3d4943adcff54a08dc0522aab9148e4,50.121353,50.121353
```